[36136] Rework members page to show name and avatars for principals

oliverguenther opened the pull request:
- [x] Output avatar for all types of principals
- [x] Remove first and lastname, add common name column that remains sortable
- [x] Extend members specs for this change 

https://community.openproject.com/wp/36136

oliverguenther committed:
Rewrite members page to output name column

oliverguenther committed:
Remove white-space nowrap for long login names in administration

ulferts commented:
@oliverguenther #8961 got merged.

oliverguenther committed:
Extend member specs

oliverguenther commented:
@ulferts I could use your help in finding out how to properly add the query order option for `name` to correctly use the full name now. There is a filter that does the column handling depending on the name formatting setting, but there's no order yet. That order would need a complex statement or an additional select to an alias that we can then sort on.

I can't seem to find a reference for how to do that other than the work packages `:sortable` definitions which IIRC was not desired either?

oliverguenther committed:
Allow query instance to be used in table cell

oliverguenther committed:
Fix memberships specs with changed columns

oliverguenther committed:
Add order for email

ulferts commented on app/cells/members/row_cell.rb:
```diff
@@ -18,12 +19,17 @@ def row_css_class
+        if user?
+          link_to principal.name, user_path(principal)
+        else
+          content_tag :span, principal.name
```
One could link this to the new group show page in case of a group.

ulferts commented on app/cells/members/table_cell.rb:
```diff
@@ -1,11 +1,11 @@
-    sortable_columns :lastname, :firstname, :mail
+    options :authorize_update, :available_roles, :is_filtered
+    columns :name, :mail, :roles, :groups, :status
+    sortable_columns :name, :mail
```
Probably not an issue as I have never heard it to be one before but it is a little bit strange that one cannot sort by status.

ulferts commented on app/models/queries/members/orders/name_order.rb:
```diff
@@ -0,0 +1,49 @@
+#++
+
+class Queries::Members::Orders::NameOrder < Queries::BaseOrder
+  self.model = Principal
```
Normally, this should be `Member` (and actually simply be deduced from the namespace). Then the `left_outer_joins` method would be overwritten to include the principal association.

ulferts reviewed (changes requested):
* The sorting by name does not work as empty values for the firstname value for groups and placeholder users lead to those principals being displayed out of order. This did exist before as well but it now becomes even more apparent.:

![image](https://user-images.githubusercontent.com/617519/107193830-90100000-69ef-11eb-806c-9649c48f477e.png)


* Just a minor thing but the "Groups" column is currently displaying the placeholder's name for a placeholder user. 

* Existing before but a good opportunity to remove it, the filter section cannot be closed: 

![image](https://user-images.githubusercontent.com/617519/107193553-41fafc80-69ef-11eb-92b5-7b2ac504e2ed.png)

Other than that, only small code remarks.

oliverguenther committed:
Review feedback

oliverguenther committed:
Allow x to close add member form

oliverguenther committed:
WIP sort firstname/lastname

oliverguenther committed:
Make status sortable

oliverguenther committed:
Extract order_by_name into the order class

oliverguenther committed:
Re-add but deprecate order_by_name

It's still in use in combination with some special scopes

(custom actions for example) that I don't want to remove right now

oliverguenther commented:
@ulferts I believe I addressed all your remarks.

ulferts commented on app/cells/members/row_cell.rb:
```diff
@@ -149,6 +144,21 @@ def column_css_class(column)
+      when Group
+        link_to principal.name, show_group_path(principal)
+      else
+        content_tag :span, principal.name
```
We need to revisit this once the placeholder users administration PR (#8944) is in.

ulferts commented on app/models/queries/users/orders/name_order.rb:
```diff
@@ -28,22 +28,8 @@
+class Queries::Users::Orders::NameOrder < Queries::Principals::Orders::NameOrder
+  # .user is important here as it forces
+  # "AND users.type = 'User'" which otherwise gets lost for some reason
+  self.model = User.user
```
This has also bugged me for reasons beyond me and is kinda unsettling.

ulferts commented on app/models/queries/principals/orders/name_order.rb:
```diff
@@ -35,15 +35,22 @@ def self.key
     end
+  end
 
-    ordered
```
This leads to the groups and placeholder users always being sorted to the end. 

An alternative would be to sort by the combined attributes:

```
Arel.sql("#{User::USER_FORMATS_STRUCTURE[Setting.user_format].join(" || ' ' || ")} #{direction = :desc ? "DESC" : "ASC"}")
```

Don't know about the performance impact, though. Maybe we need a couple of indices in the long run but I'd be willing to try it out without them for now.

oliverguenther replied:
I did something similar to the user name filter now and used `concat_ws`. I'm quite happy with the solution now

ulferts replied:
Works splendidly, thanks.

ulferts commented on app/models/principal.rb:
```diff
@@ -86,8 +86,9 @@ def self.search_scope_without_project(project, query)
-    order(User::USER_FORMATS_STRUCTURE[Setting.user_format].map { |format| "#{Principal.table_name}.#{format}" })
+  def self.order_by_name(desc: false)
+    OpenProject::Deprecation.warn "Use Queries::Principals with .order('name') instead"
+    order Queries::Principals::Orders::NameOrder.name_order_statements(desc)
```
I am not yet decided on all of this and am very open for discussion. Currently I'd rather place more of the functionality in scopes again and use them in the query order objects. 

Scopes seem to be more easy to use throughout the code. This of course is highly subjective.

oliverguenther replied:
I replaced it with a separate scope

ulferts replied:
Merci

ulferts reviewed (changes requested):
Sorry to be nagging.

oliverguenther committed:
Rewrite order_by_name into scope

oliverguenther commented:
I believe I addressed all remarks that we can right now (Placeholder PR is not yet in)
